src/AI/AICombatUtils.java

```diff
@@ -0,0 +1,274 @@
+          targetLocs.addAll(locations);
+        }
+      } // ~Weapon loop
+    }
+    targetLocs.remove(start); // No attacking your own position.
+    return targetLocs;
+  }
+
+  /**
+   * Finds a kill on the designated unit, if available.
+   * @return The lethal combination of units organized by strike location, null on failure.
+   */
+  public static HashMap<XYCoord, Unit> findAssaultKill(
+                                   GameMap gameMap, Unit target,
+                                   Commander co, Collection<Unit> attackCandidates,
+                                   Collection<XYCoord> excludedSpaces)
+  {
+    if( target.getHP() < 1 ) // Try not to pick fights with zombies
+      return null;
+
+    int minRange = 1;
+    ArrayList<XYCoord> coordsToCheck =
+        Utils.findLocationsInRange(gameMap,
+                                   new XYCoord(target.x, target.y),
+                                   minRange, findMaxStrikeWeaponRange(co));
+
+    HashMap<XYCoord, Unit> neededAttacks = new HashMap<XYCoord, Unit>();
+    // Figure out where we can attack from, and include attackers already in range by default.
+    for( XYCoord xyc : coordsToCheck )
+    {
+      if( gameMap.isLocationFogged(xyc) )
+        continue;
+      if( excludedSpaces.contains(xyc) )
+        continue;
+
+      Location loc = gameMap.getLocation(xyc);
+      Unit resident = loc.getResident();
+
+      if( null == resident || (resident.CO == co && !resident.isTurnOver) )
+        neededAttacks.put(xyc, null);
+    }
+
+    double damage = findAssaultKill(gameMap, attackCandidates, neededAttacks, target, 0);
+    if( damage >= target.getHP() )
+    {
+      // Prune excess attacks and empty attacking spaces
+      for( XYCoord space : new ArrayList<XYCoord>(neededAttacks.keySet()) )
+      {
+        Unit attacker = neededAttacks.get(space);
+        if( null == attacker )
+        {
+          neededAttacks.remove(space);
+          continue;
+        }
+        double thisShot =
+            CombatEngine.simulateBattleResults(attacker, target, gameMap,
+                                               space.xCoord, space.yCoord).defenderHPLoss;
+        if( target.getHP() <= damage - thisShot )
+        {
+          neededAttacks.remove(space);
+          damage -= thisShot;
+        }
+      }
+
+      return neededAttacks;
+    }
+
+    return null;
+  }
+
+  /**
+   * Attempts to find a combination of attacks that will create a kill.
+   * Recursive.
+   * @param unitQueue The set of potential attackers
+   * @param neededAttacks The set of locations to consider, pre-populated with any mandatory attacks, to be populated
+   * @param pDamage The cumulative base damage done by those mandatory attacks
+   * @return The cumulative base damage of all attacks in the neededAttacks
+   */
+  public static double findAssaultKill(GameMap gameMap, Collection<Unit> unitQueue, Map<XYCoord, Unit> neededAttacks, Unit target, double pDamage)
```

Comment:
I can see merit in breaking some of these functions into a separate file, but I'm not sure `findAssaultKill` really fits into the "utility" category. It contains logic specific to Wally's operation - there isn't necessarily one "right" answer, so it's not the kind of question I would normally put in a utility class.

Reply:
Why not? It finds the cheapest combination of units to clear a particular tile. There is a huge quantity of applications for that.

Reply:
Specifically the cheapest combination of units that will do the job? This should be clarified in the docstring (along with any other assumptions/requirements/context that is needed to use this properly).
Also not sure that `findAssaultKill` is really the best name for this, but eh.

Reply:
I was trying to make "Assault" and "Siege" terms for "units that can attack after moving" and "can't", since (in)direct is specifically about attack range.
You're right, though; it now happily uses indirects.

Reply:
Even aside from the fact that the function can use indirects... You can't just assign precise, domain-specific meaning to terms and then expect everyone to understand, especially when you don't record your definitions anywhere. You seem to do this often though, so I'm going to keep pointing it out.
It's one thing if you do it inside an AI, and that's the only place the term matters (though I'd still say you should treat it as public code as far as style goes); it's quite another if you are putting it in a utility for the world to misunderstand.

Reply:
> You seem to do this often though, so I'm going to keep pointing it out.

Fair enough.
I suppose "(im)mobile weapon" works for most use cases, and is more clear.

Not having a concise term for "move and fire indirect" bugs me, though.
"Mobile indirect" doesn't work, for obvious reasons.
"Assault indirect" seems like a reasonable alternative to me, and "siege unit/weapon" is pretty obvious on its own.

> it's quite another if you are putting it in a utility for the world to misunderstand.

I suppose that's why we have code review, isn't it?